### PR TITLE
make mdb2 use the actual git binary, make doing multiple jobs a bit less stupid

### PR DESCRIPTION
**Pull request by jupyterkat**

### CODE REVIEW

File: crates/mapdiffbot2/src/job_processor.rs [hunk 1]
```diff
@@ -361,34 +261,27 @@ pub fn do_job(job: &Job) -> Result<CheckOutputs> {
...
-
     let filter_on_status = |status: FileDiffStatus| {
         job.files
             .iter()
             .filter(|f| f.status == status)
             .collect::<Vec<&FileDiff>>()
     };
 
+    let diffs = git2::Diff::from_buffer(job.patch.as_ref().unwrap().as_bytes())
+        .map_err(|_| anyhow::anyhow!("Failed to parse patch, probably corrupted"))?;
```
Comment (tigercat2000) [inline crates/mapdiffbot2/src/job_processor.rs:272]: Does this work here?
```suggestion
        .context("Failed to parse patch, probably corrupted")?;
```

File: .gitignore [hunk 2]
```diff
@@ -1 +1,2 @@
 target/
+.vscode/
```
Comment (tigercat2000) [inline .gitignore:2]: Use `\.vscode` to actually get rid of the vscode directory

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (tigercat2000) [inline .vscode/launch.json:1]: This should be gitignored, see comment on the \.gitignore

File: crates/diffbot_lib/src/github/github_api.rs [hunk 4]
```diff
@@ -221,7 +212,7 @@ pub async fn download_file<S: AsRef<str>>(
     path.push(&target.sha);
     path.set_extension("dmi");
 
-    tokio::fs::create_dir_all(path.parent().unwrap()).await?;
+    async_fs::create_dir_all(path.parent().unwrap()).await?;
```
Comment (tigercat2000) [inline crates/diffbot_lib/src/github/github_api.rs:215]: Why do you use this instead of tokio::fs?

File: crates/diffbot_lib/src/github/graphql.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                    FileDiff {
+                        status,
+                        filename: item.node.path,
+                    }
+                }),
+        );
+    }
+    Ok(ret)
+}
+/*
```
Comment (tigercat2000) [inline crates/diffbot_lib/src/github/graphql.rs:160]: Please document what this commented code is useful for, or remove it~

File: crates/diffbot_lib/src/job/runner.rs [hunk 6]
```diff
@@ -1,152 +1,52 @@
...
-            let res = job.check_run.mark_succeeded(output).await;
-            if res.is_err() {
-                let _ = job
-                    .check_run
-                    .mark_failed(&format!("Failed to upload job output: {:?}", res))
-                    .await;
-            }
+pub async fn handle_output<S: AsRef<str>>(output: Vec<Output>, check_run: CheckRun, name: S) {
+    match output.len() {
+        0usize => {
```
Comment (tigercat2000) [inline crates/diffbot_lib/src/job/runner.rs:5]: .. is this suffix actually necessary? I thought rust automatically converted number literals to the relevant type

File: crates/diffbot_lib/src/job/runner.rs [hunk 7]
```diff
@@ -1,152 +1,52 @@
...
-                    .await
-                {
-                    let res = check.mark_succeeded(overflow).await;
-                    if res.is_err() {
-                        let _ = job
-                            .check_run
-                            .mark_failed(&format!("Failed to upload job output: {:?}", res))
+        }
+        _ => {
+            let len = output.len();
```
Comment (tigercat2000) [inline crates/diffbot_lib/src/job/runner.rs:19]: you can bind to output.len() in the match case, replace `_ => { let len // ...` with `len => `

File: crates/icondiffbot2/src/github_processor.rs [hunk 8]
```diff
@@ -74,17 +74,49 @@ async fn handle_pull_request(
         return Ok(());
     }
 
-    let files = get_pull_files(&payload.installation, &payload.pull_request).await?;
+    let (blacklist, contact) = {
+        let conf = &crate::CONFIG.get().unwrap();
```
Comment (tigercat2000) [inline crates/icondiffbot2/src/github_processor.rs:78]: This feels like something that should maybe be a helper?

File: crates/icondiffbot2/src/job_processor.rs [hunk 9]
```diff
@@ -3,62 +3,66 @@ use crate::{
...
-    collections::hash_map::DefaultHasher,
     fs::File,
     hash::{Hash, Hasher},
     io::{BufWriter, Write},
     path::Path,
 };
-use tokio::runtime::Handle;
-use tracing::{info_span, Instrument};
 
 #[tracing::instrument]
```
Comment (tigercat2000) [inline crates/icondiffbot2/src/job_processor.rs:21]: This works without the Instrument trait imported?

File: crates/icondiffbot2/src/job_processor.rs [hunk 10]
```diff
@@ -3,62 +3,66 @@ use crate::{
...
     job: &Job,
     diff: (Option<IconFileWithName>, Option<IconFileWithName>),
-) -> Result<(String, Vec<String>)> {
+) -> Result<(&'static str, Vec<String>)> {
     // TODO: Alphabetize
     // TODO: Test more edge cases
     match diff {
-        (None, None) => unreachable!("Diffing (None, None) makes no sense"),
+        (None, None) => Ok((
+            "UNCHANGED",
```
Comment (tigercat2000) [inline crates/icondiffbot2/src/job_processor.rs:49]: Technically this should be "moved without changes"

File: crates/icondiffbot2/src/main.rs [hunk 11]
```diff
@@ -1,30 +1,26 @@
...
 mod job_processor;
+mod runner;
 mod sha;
 mod table_builder;
 
-use diffbot_lib::job::{
-    runner::handle_jobs,
-    types::{JobJournal, JobSender},
-};
+use async_mutex::Mutex;
```
Comment (tigercat2000) [inline crates/icondiffbot2/src/main.rs:7]: Question again, why async_mutex instead of tokio's mutex?

File: crates/icondiffbot2/src/main.rs [hunk 12]
```diff
@@ -50,13 +46,16 @@ pub struct WebConfig {
 pub struct Config {
     pub github: GithubConfig,
     pub web: WebConfig,
+    pub blacklist: std::collections::HashSet<u64>,
```
Comment (tigercat2000) [inline crates/icondiffbot2/src/main.rs:49]: using a hashset for this seems like overkill and I love it lol

File: crates/icondiffbot2/src/main.rs [hunk 13]
```diff
@@ -95,14 +94,21 @@ fn read_key(path: &Path) -> Vec<u8> {
     key
 }
 
-#[tokio::main]
+const JOB_JOURNAL_LOCATION: &str = "jobs";
+
+#[actix_web::main]
 async fn main() -> std::io::Result<()> {
+    std::env::set_current_dir(std::env::current_exe().unwrap().parent().unwrap()).unwrap();
```
Comment (tigercat2000) [inline crates/icondiffbot2/src/main.rs:101]: Iiii dunno if this is a good idea... This seems like something that people running it should figure out.. this would prevent you from, say, setting this up as a systemd service with a job journal in a common location

File: crates/icondiffbot_test_generator/Cargo.toml [hunk 14]
```diff
@@ -7,7 +7,6 @@ license = "MIT"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
```
Comment (tigercat2000) [inline crates/icondiffbot_test_generator/Cargo.toml:9]: I should really finish this at some point...

File: crates/mapdiffbot2/src/github_processor.rs [hunk 15]
```diff
@@ -1,37 +1,43 @@
...
-use rocket::Request;
-use rocket::State;
-
-use crate::CONFIG;
-use diffbot_lib::github::github_api::*;
-use diffbot_lib::github::github_types::*;
-use diffbot_lib::job::types::{Job, JobJournal, JobSender};
+
+use rocket::{
+    http::Status, outcome::Outcome, request, request::FromRequest, tokio::sync::Mutex, Request,
```
Comment (tigercat2000) [inline crates/mapdiffbot2/src/github_processor.rs:6]: Should MDB also be using async_mutex instead of rocket::tokio?

File: crates/mapdiffbot2/src/github_processor.rs [hunk 16]
```diff
@@ -92,8 +105,11 @@ async fn process_pull(
         installation: InstallationId(installation.id),
     };
 
-    journal.lock().await.add_job(job.clone()).await;
-    job_sender.0.send_async(job).await?;
+    let job = serde_json::to_vec(&JobType::GithubJob(job))?;
+
+    job_sender.lock().await.send(job).await?;
```
Comment (tigercat2000) [inline crates/mapdiffbot2/src/github_processor.rs:110]: are multiple awaits like this okay?

File: crates/mapdiffbot2/src/job_processor.rs [hunk 17]
```diff
@@ -102,136 +72,98 @@ fn render(
...
-                .context("Rendering removed maps")?;
-
-                Ok(maps)
-            },
-        );
-        results.0?;
-        results.1
-        //eprintln!("Base maps took {}ms", now.elapsed().as_millis());
+    let removed_maps = with_checkout_and_dir(&base_branch, repo, &path, || {
+        render_map_regions(
```
Comment (tigercat2000) [inline crates/mapdiffbot2/src/job_processor.rs:107]: This is no longer multithreaded? Is that okay?

**Review by tigercat2000 — CHANGES_REQUESTED** [review]

**Comment by tigercat2000** [comment]
I'll leave a note here instead of in the review- we should really settle on one web framework used for both MDB and IDB lol
I think it should be actix_web because Rocket interferes with tracing because it does a bunch of automatic console setup

File: .vscode/launch.json (hunk 3 above)
Comment (ZeWaka) [reply .vscode/launch.json:1]: Should it be? This seems helpful.

File: crates/mapdiffbot2/src/runner.rs [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+
+async fn garbage_collect_all_repos() {
+    use eyre::Result;
+    use path_absolutize::Absolutize;
+    use std::process::Command;
+    info!("Garbage collection starting!");
+
+    let output = rocket::tokio::time::timeout(
+        Duration::from_secs(3600),
+        //tfw no try blocks
```
Comment (ZeWaka) [inline crates/mapdiffbot2/src/runner.rs:39]: rip `try!` 😭

File: crates/mapdiffbot2/src/github_processor.rs (hunk 16 above)
Comment (ZeWaka) [reply crates/mapdiffbot2/src/github_processor.rs:110]: yes

File: crates/icondiffbot2/src/main.rs (hunk 11 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/main.rs:7]: idb uses actix's runtime now, it doesn't include a mutex

File: crates/icondiffbot2/src/main.rs (hunk 13 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/main.rs:101]: they shouldn't be, or weird things will happen

File: crates/icondiffbot2/src/main.rs (hunk 11 above)
Comment (tigercat2000) [reply crates/icondiffbot2/src/main.rs:7]: wild

File: crates/icondiffbot2/src/main.rs (hunk 13 above)
Comment (tigercat2000) [reply crates/icondiffbot2/src/main.rs:101]: oh? what kind of weird things?

Comment (jupyterkat) [reply crates/icondiffbot2/src/main.rs:101]: the app changes cwd regularly, and yaque will fail to delete the lock file if it can't find the job journal with relative paths

Comment (jupyterkat) [reply crates/icondiffbot2/src/main.rs:101]: this cwd-based pathing is awful, ill probably redo it sometime later

File: crates/diffbot_lib/src/job/runner.rs (hunk 6 above)
Comment (jupyterkat) [reply crates/diffbot_lib/src/job/runner.rs:5]: it's interpreted as u32 the last time i tried

File: crates/icondiffbot2/src/job_processor.rs (hunk 10 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/job_processor.rs:49]: this condition shouldn't be hit at all, i've simply made it not panic

File: crates/mapdiffbot2/src/job_processor.rs (hunk 17 above)
Comment (jupyterkat) [reply crates/mapdiffbot2/src/job_processor.rs:107]: yes

File: .vscode/launch.json (hunk 3 above)
Comment (tigercat2000) [reply .vscode/launch.json:1]: it does but they also tried to get rid of it so

Comment (jupyterkat) [reply .vscode/launch.json:1]: no, i wanted to provide a default config for people, and one that they can change to their own without git getting in the way

File: crates/diffbot_lib/src/github/github_api.rs (hunk 4 above)
Comment (jupyterkat) [reply crates/diffbot_lib/src/github/github_api.rs:215]: shared lib, don't want another tokio

File: crates/diffbot_lib/src/github/graphql.rs (hunk 5 above)
Comment (jupyterkat) [reply crates/diffbot_lib/src/github/graphql.rs:160]: in progress code for a graphql mutation on checkrun, figured it's useless now

File: crates/mapdiffbot2/src/github_processor.rs (hunk 15 above)
Comment (jupyterkat) [reply crates/mapdiffbot2/src/github_processor.rs:6]: it can use it's own mutex because it's there

File: crates/icondiffbot2/src/github_processor.rs (hunk 8 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/github_processor.rs:78]: why does getting a config entry need to be a helper

File: crates/icondiffbot2/src/main.rs [hunk 19]
```diff
@@ -98,8 +98,6 @@ const JOB_JOURNAL_LOCATION: &str = "jobs";
 
 #[actix_web::main]
 async fn main() -> std::io::Result<()> {
-    std::env::set_current_dir(std::env::current_exe().unwrap().parent().unwrap()).unwrap();
-
     stable_eyre::install().expect("Eyre handler installation failed!");
```
Comment (AffectedArc07) [inline crates/icondiffbot2/src/main.rs:101]: Yeah I dont think this is a good idea

The whole reason we did this check is because of the directory swapping with sdmm and other stuff so much, and it kept making repo clones inside other repos and all sorts of unpleasant business.

I can try running this but I seem to remember needing a ton of path sets for this to be safe.

File: crates/icondiffbot2/src/main.rs (hunk 13 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/main.rs]: redone

File: crates/icondiffbot2/src/main.rs (hunk 19 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/main.rs:101]: it doesn't set the cwd willy nilly anymore, so this should be fine

Comment (AffectedArc07) [reply crates/icondiffbot2/src/main.rs:101]: > it doesn't set the cwd willy nilly anymore, so this should be fine

Do you have this much faith in sdmm?

File: .vscode/launch.json (hunk 3 above)
Comment (tigercat2000) [reply .vscode/launch.json:1]: @jupyterkat make the gitignore `\.vscode` and `!\.vscode/launch.json` then

File: crates/mapdiffbot2/src/github_processor.rs (hunk 15 above)
Comment (tigercat2000) [reply crates/mapdiffbot2/src/github_processor.rs:6]: hmm, something to do if we yeet rocket, mkay

File: crates/icondiffbot2/src/job_processor.rs (hunk 10 above)
Comment (tigercat2000) [reply crates/icondiffbot2/src/job_processor.rs:49]: I was wrong about making this unreachable lol, it happens in normal cases of "we moved a file without changes"

File: crates/icondiffbot2/src/github_processor.rs (hunk 8 above)
Comment (tigercat2000) [reply crates/icondiffbot2/src/github_processor.rs:78]: it uses an internal block, s' a little strange

File: crates/icondiffbot2/src/job_processor.rs (hunk 10 above)
Comment (jupyterkat) [reply crates/icondiffbot2/src/job_processor.rs:49]: no, it's impossible to reach now, i've filtered out all the cases that lead to this

Comment (tigercat2000) [reply crates/icondiffbot2/src/job_processor.rs:49]: ChangeType::Renamed isn't actually consistently reported by github, it can give you a moved-without-changes as a ChangeType::Modified

File: .vscode/launch.json (hunk 3 above)
Comment (jupyterkat) [reply .vscode/launch.json:1]: you don't get the point, do ya, i don't want git to track launch.json

Comment (jupyterkat) [reply .vscode/launch.json:1]: but i do want it to CREATE launch.json

**Review by tigercat2000 — APPROVED** [review]
its good

**Comment by AffectedArc07** [comment]
Will merge once coffee fixes the clippy lint, and its ran fine in prod.

Then I can begin the deprecation of MDB1 and IDB2. Bless the work of all 3 of you.

**Comment by AffectedArc07** [comment]
If the GC runs fine tomorrow morning, I am hitting merge on this and retiring the old 2.

**Comment by AffectedArc07** [comment]
```
Dec 30 04:00:00 AA07-S-UBUNTU start.sh[523505]: 04:00:00 [INFO] async_spawn_by_tokio: routine-exec; task_id=1 record_id=7014439927719006208
Dec 30 04:00:00 AA07-S-UBUNTU start.sh[523505]: 04:00:00 [INFO] Job received from queue
Dec 30 04:00:00 AA07-S-UBUNTU start.sh[523505]: 04:00:00 [INFO] Garbage collection starting!
Dec 30 04:14:58 AA07-S-UBUNTU start.sh[523505]: 04:14:58 [INFO] Garbage collection finished!
```

It worky!

**Comment by AffectedArc07** [comment]
WE GAMING NOW
